### PR TITLE
wip to refactor methodprovidercollection

### PR DESCRIPTION
**Pull request by m-nash**
contributes to https://github.com/microsoft/typespec/issues/3213

### CODE REVIEW

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/ScmMethodProviderCollection.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                methodModifier |= MethodSignatureModifiers.Async;
+            }
+            var opName = operation.Name.ToCleanName();
+            var methodSignature = new MethodSignature(
+                isAsync ? opName + "Async" : opName,
+                FormattableStringHelpers.FromString(operation.Description),
+                methodModifier,
+                GetResponseType(operation.Responses, isAsync),
+                null,
+                Parameters: [.. methodParameters, KnownParameters.CancellationTokenParameter]);
```
Comment (ShivangiReja) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/ScmMethodProviderCollection.cs:55]: We take RequestOptions parameter in protocol methods right?

File: packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs [hunk 2]
```diff
@@ -15,89 +17,435 @@ internal void CreateSayHiRequest(Uri unbrandedTypeSpecUrl, string headParameter,
...
             throw new NotImplementedException("Method not implemented.");
         }
 
+        /// <summary> Return hi again. </summary>
+        /// <param name="p1"></param>
+        /// <param name="p2"></param>
+        /// <param name="action"></param>
+        /// <param name="cancellationToken"> The cancellation token to use. </param>
+        /// <exception cref="ArgumentNullException"> <paramref name="p1"/>, <paramref name="p2"/> or <paramref name="action"/> is null. </exception>
+        public virtual void HelloAgain(string p1, string p2, RoundTripModel action, CancellationToken cancellationToken = default)
```
Comment (ShivangiReja) [inline packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs:58]: It looks like this is a convenience method. I see that we don't take a CancellationToken parameter in the typespec-generated code.

Is this PR just to refactor MethodProviderCollection, and will you have a separate PR to generate protocol methods?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/ScmMethodProviderCollection.cs (hunk 1 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/ScmMethodProviderCollection.cs:55]: this is just a wip and will be fixed before the issue closes but allows others to work off of this refactor

File: packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs (hunk 2 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/TestProjects/Local/Unbranded-TypeSpec/src/Generated/UnbrandedTypeSpecClient.cs:58]: this is just a wip and will be fixed before the issue closes but allows others to work off of this refactor

**Review by ShivangiReja — APPROVED** [review]

**Pull request merged**